File: wix-restaurants-authentication-client/src/main/java/com/wix/restaurants/authentication/DefaultWixRestaurantsAuthenticationClient.java
package com.wix.restaurants.authentication;

import com.fasterxml.jackson.core.type.TypeReference;
import com.google.api.client.http.HttpRequestFactory;
import com.google.api.client.http.javanet.NetHttpTransport;
import com.wix.restaurants.authentication.exceptions.ExceptionsTranslator;
import com.wix.restaurants.authentication.exceptions.InternalException;
import com.wix.restaurants.authentication.model.*;
import com.wix.restaurants.jsonclient.JsonClient;

import java.io.IOException;

public class DefaultWixRestaurantsAuthenticationClient implements WixRestaurantsAuthenticationClient {
    private final ExceptionsTranslator exceptionsTranslator = new ExceptionsTranslator();
    private final JsonClient jsonClient;
    private final String endpointUrl;

    public static class Builder {
        private HttpRequestFactory requestFactory = new NetHttpTransport().createRequestFactory();
        private int connectTimeout = 0;
        private int readTimeout = 0;
        private int numberOfRetries = 0;
        private String endpointUrl = Endpoints.PRODUCTION;

        public Builder setRequestFactory(HttpRequestFactory requestFactory) {
            this.requestFactory = requestFactory;
            return this;
        }

        public Builder setConnectTimeout(int connectTimeout) {
            this.connectTimeout = connectTimeout;
            return this;
        }

        public Builder setReadTimeout(int readTimeout) {
            this.readTimeout = readTimeout;
            return this;
        }

        public Builder setNumberOfRetries(int numberOfRetries) {
            this.numberOfRetries = numberOfRetries;
            return this;
        }

        public Builder setEndpointUrl(String endpointUrl) {
            this.endpointUrl = endpointUrl;
            return this;
        }

        public DefaultWixRestaurantsAuthenticationClient build() {
            return new DefaultWixRestaurantsAuthenticationClient(
                    requestFactory, connectTimeout, readTimeout, numberOfRetries, endpointUrl);
        }
    }

    public DefaultWixRestaurantsAuthenticationClient(HttpRequestFactory requestFactory, Integer connectTimeout,
                                                     Integer readTimeout, Integer numberOfRetries, String endpointUrl) {
        jsonClient = new JsonClient(requestFactory, connectTimeout, readTimeout, numberOfRetries);
        this.endpointUrl = endpointUrl;
    }

    @Override
    public LoginResponse loginWithFacebook(String fbAccessToken) {
        final FacebookLoginRequest facebookLoginRequest = new FacebookLoginRequest();
        facebookLoginRequest.accessToken = fbAccessToken;

        final LoginResponse facebookLoginResponse = request(
                facebookLoginRequest, new TypeReference<Response<LoginResponse>>() {});

        return facebookLoginResponse;
    }

    @Override
    public LoginResponse loginWithOpenrest(String username, String password) {
        final OpenrestLoginRequest openrestLoginRequest = new OpenrestLoginRequest();
        openrestLoginRequest.username = username;
        openrestLoginRequest.password = password;

        final LoginResponse openrestLoginResponse = request(
                openrestLoginRequest, new TypeReference<Response<LoginResponse>>() {});

        return openrestLoginResponse;
    }

    @Override
    public LoginResponse loginWithWixInstance(String instance) {
        final WixLoginInstanceRequest wixLoginInstanceRequest = new WixLoginInstanceRequest();
        wixLoginInstanceRequest.instance = instance;

        final LoginResponse wixLoginInstanceResponse = request(
                wixLoginInstanceRequest, new TypeReference<Response<LoginResponse>>() {});

        return wixLoginInstanceResponse;
    }

    @Override
    public User authenticate(String accessToken) {
        final AuthenticateRequest authenticateRequest = new AuthenticateRequest();
        authenticateRequest.accessToken = accessToken;

        final User authenticateResponse = request(
                authenticateRequest, new TypeReference<Response<User>>() {});

        return authenticateResponse;
    }

    private <T> T request(Request request, TypeReference<Response<T>> responseType) {
        try {
            final Response<T> response = jsonClient.post(endpointUrl, request, responseType);

            if (response.error != null) {
                throw exceptionsTranslator.translateError(response.error);
            }

            return response.value;
        } catch (IOException e) {
            throw new InternalException(e.getMessage(), e);
        }
    }
}

File: wix-restaurants-authentication-client/src/main/java/com/wix/restaurants/authentication/DefaultWixRestaurantsAuthenticationClient.scala
package com.wix.restaurants.authentication

import akka.actor.ActorSystem
import akka.http.javadsl.model.headers.Authorization
import akka.http.scaladsl.client.RequestBuilding.{Get, Post}
import com.wix.rest.rfc7807.api.model.ErrorResponse
import com.wix.rest.rfc7807.client.AkkaRestClient
import com.wix.restaurants.authentication.exceptions._
import com.wix.restaurants.authentication.model._
import com.wix.restaurants.json.Json

import scala.concurrent.duration.Duration
import scala.concurrent.{Await, ExecutionContext}

class DefaultWixRestaurantsAuthenticationClient(endpoint: String = "https://auth.wixrestaurants.com/v2",
                                                readTimeout: Option[Duration] = None) extends WixRestaurantsAuthenticationClient {
  private val actualReadTimeout: Duration = readTimeout.getOrElse(Duration.Inf)

  private implicit val system: ActorSystem = ActorSystem("akka-wix-restaurants-client-system")
  private implicit val executionContext: ExecutionContext = system.dispatcher
  private val client: AkkaRestClient = new AkkaRestClient(errorResponseAsException = ExceptionTranslator.asException)


  override def loginWithFacebook(fbAccessToken: String): LoginResponse = {
    val request = Post(s"$endpoint/com.facebook/access_token", Json.stringify(FacebookLogin(fbAccessToken)))
    Await.result(client.execute(request) withResult[LoginResponse](), actualReadTimeout)
  }

  override def loginWithGoogle(clientId: String, idToken: String): LoginResponse = {
    val request = Post(s"$endpoint/com.google/access_token", Json.stringify(GoogleLogin(clientId, idToken)))
    Await.result(client.execute(request) withResult[LoginResponse](), actualReadTimeout)
  }

  override def loginWithOpenrest(username: String, password: String): LoginResponse = {
    val request = Post(s"$endpoint/com.openrest/access_token", Json.stringify(OpenrestLogin(username, password)))
    Await.result(client.execute(request) withResult[LoginResponse](), actualReadTimeout)
  }

  override def loginWithWixInstance(instance: String): LoginResponse = {
    val request = Post(s"$endpoint/com.wix/access_token", Json.stringify(WixLogin(instance)))
    Await.result(client.execute(request) withResult[LoginResponse](), actualReadTimeout)
  }

  override def loginInstanceWithWixInstance(instance: String): LoginResponse = {
    val request = Post(s"$endpoint/com.wix.instances/access_token", Json.stringify(WixLogin(instance)))
    Await.result(client.execute(request) withResult[LoginResponse](), actualReadTimeout)
  }

  override def impersonate(accessToken: String, user: User): LoginResponse = {
    val request = Post(s"$endpoint/impersonate", Json.stringify(user))
      .addHeader(Authorization.oauth2(accessToken))
    Await.result(client.execute(request) withResult[LoginResponse](), actualReadTimeout)
  }

  override def authenticate(accessToken: String): User = {
    val request = Get(s"$endpoint/me")
      .addHeader(Authorization.oauth2(accessToken))
    Await.result(client.execute(request) withResult[User](), actualReadTimeout)
  }
}

private object ExceptionTranslator {
  def asException(errorResponse: ErrorResponse): RuntimeException = {
    // TODO: use error type constants
    errorResponse.`type` match {
      case "https://www.wixrestaurants.com/errors/internal" => new InternalException(errorResponse.detail.orNull)
      case "https://www.wixrestaurants.com/errors/authentication" => new AuthenticationException(errorResponse.detail.orNull)
      case "https://www.wixrestaurants.com/errors/invalid_data" => new InvalidDataException(errorResponse.detail.orNull)
      case "https://www.wixrestaurants.com/errors/forbidden" => new ForbiddenException(errorResponse.detail.orNull)
      case "https://www.wixrestaurants.com/errors/not_secure" => new SecurityException(errorResponse.detail.orNull)
      case _ => new RuntimeException(s"Type: '${errorResponse.`type`}'${errorResponse.detail.map(detail => s", Detail: '$detail'").getOrElse("")}")
    }
  }
}
